server/src/main/java/org/opensearch/search/SearchService.java

```diff
@@ -1081,24 +1098,23 @@ private DefaultSearchContext createSearchContext(ReaderContext reader, ShardSear
             if (request.source() != null
                 && request.source().size() != 0
                 && (request.source().getDerivedFieldsObject() != null || request.source().getDerivedFields() != null)) {
-                Map<String, MappedFieldType> derivedFieldTypeMap = new HashMap<>();
-                if (request.source().getDerivedFieldsObject() != null) {
-                    Map<String, Object> derivedFieldObject = new HashMap<>();
-                    derivedFieldObject.put(DerivedFieldMapper.CONTENT_TYPE, request.source().getDerivedFieldsObject());
-                    derivedFieldTypeMap.putAll(
-                        DerivedFieldMapper.getAllDerivedFieldTypeFromObject(derivedFieldObject, searchContext.mapperService())
-                    );
-                }
-                if (request.source().getDerivedFields() != null) {
-                    for (DerivedField derivedField : request.source().getDerivedFields()) {
-                        derivedFieldTypeMap.put(
-                            derivedField.getName(),
-                            DerivedFieldMapper.getDerivedFieldType(derivedField, searchContext.mapperService())
-                        );
-                    }
-                }
-                context.setDerivedFieldTypes(derivedFieldTypeMap);
-                searchContext.getQueryShardContext().setDerivedFieldTypes(derivedFieldTypeMap);
+                DerivedFieldResolver derivedFieldResolver = DerivedFieldResolverFactory.createResolver(
+                    searchContext.getQueryShardContext(),
+                    request.source().getDerivedFieldsObject(),
+                    request.source().getDerivedFields(),
+                    context.getIndexSettings().isDerivedFieldAllowed() && allowDerivedField
+                );
+                context.setDerivedFieldResolver(derivedFieldResolver);
+                searchContext.getQueryShardContext().setDerivedFieldResolver(derivedFieldResolver);
+            } else {
+                DerivedFieldResolver derivedFieldResolver = DerivedFieldResolverFactory.createResolver(
+                    searchContext.getQueryShardContext(),
+                    emptyMap(),
+                    emptyList(),
+                    context.getIndexSettings().isDerivedFieldAllowed() && allowDerivedField
+                );
+                context.setDerivedFieldResolver(derivedFieldResolver);
+                searchContext.getQueryShardContext().setDerivedFieldResolver(derivedFieldResolver);
             }
```

Comment:
Can you reduce the duplication here with something like:

```
DerivedFieldResolver derivedFieldResolver = DerivedFieldResolverFactory.createResolver(
  searchContext.getQueryShardContext(),
  Objects.requireNonNullElse(request.source().getDerivedFieldsObject(), emptyMap()),
  Objects.requireNonNullElse(request.source().getDerivedFields(), emptyList()),
  context.getIndexSettings().isDerivedFieldAllowed() && allowDerivedField
);
```

Reply:
done